sw/device/lib/common.h

```diff
@@ -32,6 +32,24 @@ static const unsigned long UART_BAUD_RATE = 230400;
 
 #define ARRAYSIZE(x) (sizeof(x) / sizeof(x[0]))
 
+#define NO_RETURN __attribute__((noreturn))
+
+// This macro expands to a C statement that will loop forever without making
+// progress, without invoking undefined behavior.
+//
+// This macro should be used to terminate NO_RETURN functions that would
+// otherwise return.
+#define LOOP_FOREVER()   \
+  do {                   \
+    asm volatile("wfi"); \
+  } while (true)
+
+// This macro expands to a C statement that will immediately raise an illegal
+// instruction; the compiler will treat it as an expression that cannot return.
+#define ILLEGAL_INSTRUCTION()       \
+  asm volatile("c.unimp; c.unimp"); \
+  LOOP_FOREVER()
```

Comment:
I think just `unimp` is probably better, and I think it's surprising that this contains an infinite loop too.

Reply:
The infinite loop is desired so that the macro will be noreturn. As far as I know, there isn't a way for me to tell GCC/Clang "hey this assembly block never returns, don't emit diagnostics because I didn't put a return after it".